File: src/main/java/planetGenerator/ui/planet/viewProviders/ShadowViewProvider.java
package planetGenerator.ui.planet.viewProviders;

import javafx.scene.canvas.GraphicsContext;
import javafx.scene.layout.Pane;
import javafx.scene.paint.Color;
import planetGenerator.model.Planet;
import rak.utility.MathFunctions;

public class ShadowViewProvider extends GraphicsContextViewProvider {
	private float shadowStrength = .75f;
	private int falloff = 5;
	
	@Override
	public void drawView(Planet planet, Pane viewBackground) {
		GraphicsContext gg = createGraphicsContext(viewBackground, planet.getSize());
		int center = planet.getSize()/2;
		
		for (int x=0; x<planet.getSize(); x++){
			for (int y=0; y<planet.getSize(); y++){
				double shadowLevel = getShadowLevel(center, x, y);
				gg.setFill(new Color(0, 0, 0, shadowLevel));
				gg.fillRect(x, y, 1, 1);
			}
		}
	}
	
	private double getShadowLevel(int center, int x, int y) {
		int radius = center; //The radius and the center are the same
		
		double distanceFromCenter = MathFunctions.distanceBetween(center, center, x, y);
		double distanceFromEdge = radius - distanceFromCenter;
		distanceFromEdge = MathFunctions.clamp(distanceFromEdge, 0, radius);
		
		double percent = 1 - (distanceFromEdge/radius);
		double adjustedPercent = shadowStrength * Math.pow(percent, falloff);
		
		return MathFunctions.clamp(adjustedPercent, 0, 1);
	}

	public void setShadowStrength(float shadowStrength) {
		this.shadowStrength = shadowStrength;
	}

	public void setFalloff(int falloff) {
		this.falloff = falloff;
	}

	public float getShadowStrength() {
		return shadowStrength;
	}

	public int getFalloff() {
		return falloff;
	}
	
	

}

File: src/main/java/rak/planetGenerator/ui/planet/viewProviders/ShadowViewProvider.kt
package rak.planetGenerator.ui.planet.viewProviders

import javafx.scene.layout.Pane
import javafx.scene.paint.Color
import rak.planetGenerator.model.Planet
import rak.utility.MathFunctions

class ShadowViewProvider : GraphicsContextViewProvider() {
    var shadowStrength = .75f
    var falloff = 5

    override fun drawView(planet: Planet, viewBackground: Pane) {
        val gg = createGraphicsContext(viewBackground, planet.size)
        val center = planet.size / 2

        for (x in 0 until planet.size) {
            for (y in 0 until planet.size) {
                val shadowLevel = getShadowLevel(center, x, y)
                gg.fill = Color(0.0, 0.0, 0.0, shadowLevel)
                gg.fillRect(x.toDouble(), y.toDouble(), 1.0, 1.0)
            }
        }
    }

    private fun getShadowLevel(center: Int, x: Int, y: Int): Double {

        val distanceFromCenter = MathFunctions.distanceBetween(center, center, x, y)
        var distanceFromEdge = center - distanceFromCenter
        distanceFromEdge = MathFunctions.clamp(distanceFromEdge, 0.0, center.toDouble())

        val percent = 1 - distanceFromEdge / center
        val adjustedPercent = shadowStrength * Math.pow(percent, falloff.toDouble())

        return MathFunctions.clamp(adjustedPercent, 0.0, 1.0)
    }


}
